Serverless

jeremy-daily opened the pull request:
An initial request to review and see the ecosystem of uploading and analyzing the CAN Logger Data to the Amazon Web Services. This PR demonstrates the successful upload and storage of the data and meta-data.

jeremydaily committed:
Added initial framework

jeremy-daily committed:
Initial User Authorization

jeremy-daily committed:
Merge branch 'master' into serverless

jeremy-daily committed:
Prepared for the use of the ATEC Chip.

jeremy-daily committed:
Completed ATECC Functions in Teensy Sketch

Used the sparkfun library to get the advanced crytpo finished. Still need to safely store the server private key.

jeremy-daily committed:
Started Readme FIle

jeremy-daily committed:
cleaned the ecdsa verification

jeremy-daily committed:
Added Example Sketches

jeremy-daily committed:
Merge pull request #3 from Heavy-Vehicle-Networking-At-U-Tulsa/serverless

Serverless

jeremy-daily committed:
Download SD cards over the USB

jeremy-daily committed:
Working AWS upload with pre-signed links

jeremy-daily committed:
Added Key and auth

jeremy-daily committed:
tested auth

jeremy-daily committed:
working decryption for session key

jeremy-daily committed:
Working provisioning

jeremy-daily committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

jeremy-daily committed:
removed username

jeremy-daily committed:
Added Decryption Routine

Need to add SHA checking to determine authenticity of encrypted data. Currently the decryption is not correct.

jeremy-daily committed:
Working Decryption

Needed to hardcode the Server Public Key.

jeremy-daily committed:
Update CANLoggerClient.py

Tuned to match response from server

jeremy-daily committed:
Successful decryption of session keys

zachaubin commented on CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:
```diff
@@ -678,6 +655,7 @@ void open_binFile(){
   else
   {
     YELLOW_LED_fast_blink_state == false;
```
"YELLOW_LED_fast_blink_state == false;" 
is boolean, 
should it be:
"if (YELLOW_LED_fast_blink_state == false) write_initial_meta_data()"; 
or
"YELLOW_LED_fast_blink_state = false;"
?

zachaubin commented on CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:
```diff
@@ -989,46 +1006,140 @@ void setup(void) {
+    memcpy(&data_file_contents[data_file_index],&hex_digit,2);
+    data_file_index+=2;
+  }
+    
```
print a newline? 
"Serial.print(hex_digit);" will leave output cursor on 'that' line when for loop concludes

zachaubin commented on CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:
```diff
@@ -989,46 +1006,140 @@ void setup(void) {
+  sha256Instance->final(hash);
+  delete sha256Instance;
+  
+  Serial.print(",TXT-SHA:");
```
maybe newline instead of comma, 
generally could also be done in other similar locations
unless the compact output is desired

jeremy-daily replied:
The original intent was to have multiple lines in one file. Now we are looking at one file per log, so the newline is acceptable. The PC and Server side utilities will need to be updated too.

zachaubin commented on CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:
```diff
@@ -989,46 +1006,140 @@ void setup(void) {
+  //Write the metadata file
+  sprintf(data_file_name,"%s%s%s.txt",brand_name,logger_name,current_file);
+  if (!dataFile.open(data_file_name, O_RDWR | O_CREAT)) {
+    YELLOW_LED_fast_blink_state == true;
```
YELLOW_LED_fast_blink_state == true;
change to 
YELLOW_LED_fast_blink_state = true;

zachaubin commented on CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:
```diff
@@ -859,7 +835,28 @@ void myLongPressFunction(){
+  
+  Serial.println("Starting CAN logger.");
 
+  // The server public key should already be installed in the ATECC. Why do we have to do this?
```
according to: 
https://www.youtube.com/watch?v=SFpkqmwoEeg&t=23m57s
it's how the configuration of the 508a/608a is currently set up
not sure if it's the OTP part, haven't gone deep with the eval yet

zachaubin commented on CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:
```diff
@@ -989,46 +1006,140 @@ void setup(void) {
+    Serial.println(data_file_name);
+  }
+  else{
+    for (int i = 0; i<strlen((const char*)data_file_contents);i++){
```
"for (int i = 0; i<strlen((const char*)data_file_contents);i++){"
will become 
"for (int i = 0; i < strlen((const char*)data_file_contents); i++){"

zachaubin commented on CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256/CAN_Logger_with_Autobaud_and_Requests_no_CAN2_AES_CBC_SHA256.ino:
```diff
@@ -1162,110 +1273,101 @@ void loop(void) {
-        commandString.remove(0,4);
-        commandString.toCharArray(delete_file_name,13);
-        delete_file(delete_file_name);
+    if      (commandString.equalsIgnoreCase("HEX"))        print_hex();
```
this whole if else thing looks like it should be a switch statement
line 1276-1373

instead of cs.equalsIgnoreCase on each, do switch(cs.toUpper()) { .. } or equivalent
if not a switch, at least convert toUpper once and then compare with each in the ifelse( )

zachaubin commented on Libraries/SparkFun_ATECCX08a_Arduino_Library/src/SparkFun_ATECCX08a_Arduino_Library.cpp:
```diff
@@ -0,0 +1,1201 @@
+    }
+    Serial.println();
+  }
+}
```
at end of this function we need
"return true;"
to avoid compiler warnings

I have an idea for bitwise anding results to verify successful reads, will ask Dr. Daily about it

zachaubin commented on Libraries/SparkFun_ATECCX08a_Arduino_Library/src/SparkFun_ATECCX08a_Arduino_Library.cpp:
```diff
@@ -0,0 +1,1201 @@
+
+
+
+
```
no extra lines

zachaubin commented on Libraries/SparkFun_ATECCX08a_Arduino_Library/src/SparkFun_ATECCX08a_Arduino_Library.h:
```diff
@@ -0,0 +1,170 @@
+// 		_ _ ? ?  ? ? _ _ 	Bits 5-2 Slot number (in this example, we use slot 0, so "0 0 0 0")
+// 		_ _ _ _  _ _ ? ? 	Bits 1-0 Zone or locktype. 00=Config, 01=Data/OTP, 10=Single Slot in Data, 11=illegal
+
+#define LOCK_MODE_ZONE_CONFIG 			0b10000000
```
if there's a reason to have the values not in a column, ok, but it looks messy
lines 65 - 84

zachaubin commented on README.md:
```diff
@@ -1,9 +1,48 @@
+  1. Reliably capture all network data when connected to a vehicle on multiple vehicle networks to include 3 CAN channels and J1708.
+  2. Securely store all the collected data in non-volitile memory. This means all the data collected will be encrypted before written to the SD card.
+  3. The encryption key will be ephemeral, which means each data set has its own key for each session. 
+  4. The ephemeral key is envelope encrypted using unique key generated for each device. If one of these keys is compromized, then only one device is compromised.
```
to
"  4. The ephemeral key is envelope encrypted using a unique key generated for each device. If one of these keys is compromised, then only one device is compromised."

zachaubin commented on README.md:
```diff
@@ -1,9 +1,48 @@
+  4. The ephemeral key is envelope encrypted using unique key generated for each device. If one of these keys is compromized, then only one device is compromised.
+  5. The keys used to decrypt the records are accountable to users. This places access controls on the data sets. 
+  6. The encrypted data and meta-data are authenticated using a secure hash algorithm (SHA-256) and ellipic-curve digital signature algorithm (ECDSA).
+  7. The project makes appropriate use of the ATECC608A ECC CryptoAuthentication module from Microchip.
```
will it also work on ATECC508A?

zachaubin commented on README.md:
```diff
@@ -1,9 +1,48 @@
+  * Cognito -  provides access to user managment and authentication
+  * S3 - Amazon's simple storage service that's used to keep the data
+
+*static-web:* the data and files needed to host on a static web service. For this project, the static site is hosted at https://www.engr.colostate.edu/~jdaily/CANLogger/
```
url >> Wild Rydes unicorns?

zachaubin commented on serverless/utils.py:
```diff
@@ -0,0 +1,124 @@
+    """
+    Validate and decode the web token from the Amazon Cognito.
+    Stores the public key needed to decrypt the token.
+    Returns 
```
Returns ... ?
False if issues, else just returns?
just looking for a bit of verbosity, the code does sort of explain itself though

zachaubin commented on tests/Teensy_ECDSA_Sign/Teensy_ECDSA_Sign.ino:
```diff
@@ -0,0 +1,148 @@
+};
+
+void setup() {
+  while(!Serial);
```
I kind of want a message here, it'll just hang
is there another way to print/err besides Serial.print ?

zachaubin commented on tests/Teensy_ECDSA_Sign/Teensy_ECDSA_Sign.ino:
```diff
@@ -0,0 +1,148 @@
+    Serial.print("0x");
+    if ((message[i] >> 4) == 0) Serial.print("0"); // print preceeding high nibble if it's zero
+    Serial.print(message[i], HEX);
+    if (i != 31) Serial.print(", ");
```
I think this can be rearranged to: 
if ((31 - i) % 16 != 0) {
Serial.print(", ");
} else {
Serial.println();
}
but it's only getting rid of one "if" check so not much would be saved.
similarly below
it does read easily as it is now, this might be premature optimization

zachaubin commented on tests/Teensy_ECDSA_Verify/Teensy_ECDSA_Verify.ino:
```diff
@@ -0,0 +1,184 @@
+// note, you will also need to change the name of the copied array 
+// "publicKey[64]" to "publicKeyExternal[64]"
+
+uint8_t publicKeyExternal[64] = {0x76, 0xE2, 0x92, 0x5C, 0xC1, 0xB1, 0xC5, 0x47, 0x20, 0x20, 0xB8, 0xCE, 0x84, 0xDD, 0x27, 0xE4, 
```
newline after bracket to match others

zachaubin commented on serverless/README.md:
````diff
@@ -0,0 +1,168 @@
+  5. Create a profile with programatic access. Open a Command Prompt (type `CMD` at the start menu). Be sure you are in your home directory and run `aws configure`. There are four pieces of data needed:
+```
+AWS Access Key ID [None]: AKIAIOSFODNN7EXAMPLE
+AWS Secret Access Key [None]: wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
````
is it okay to put this here in plaintext?

zachaubin commented on serverless/README.md:
````diff
@@ -0,0 +1,168 @@
+aws_secret_access_key = **************************************RA
+```
+
+  5. Install NodeJS. https://nodejs.org/en/ (Note: installing the extra chocolaty programs take a long time.)
````
*Chocolatey

zachaubin commented on clientApp/CANLoggerClient.py:
```diff
@@ -0,0 +1,664 @@
+
+
+    def connect_logger_by_usb(self):
+        items =[] 
```
items = []

zachaubin commented on clientApp/CANLoggerClient.py:
```diff
@@ -0,0 +1,664 @@
+            character = self.serial_queue.get()
+            ret_val += character
+        
+        response=ret_val.split(b'\r\n')
```
response = ret_val.split(b'\r\n')

zachaubin commented on clientApp/CANLoggerClient.py:
```diff
@@ -0,0 +1,664 @@
+            #send the command to the Logger to download data
+            self.ser.write(b'BIN ' + bytes(latest_file_name[:-3]+"txt",'ascii') + b'\n') 
+            time.sleep(0.1);
+            ret_val=b''
```
ret_val = b''

zachaubin commented on clientApp/CANLoggerClient.py:
```diff
@@ -0,0 +1,664 @@
+                    except IndexError:
+                        d = entry #There was no colon, so 
+                    self.device_file_table.setItem(row,col,QTableWidgetItem(d))
+                    col+=1
```
col += 1
row += 1

zachaubin commented on clientApp/CANLoggerClient.py:
```diff
@@ -0,0 +1,664 @@
+            }
+        url = "https://cognito-idp.us-east-2.amazonaws.com"
+        header = {}
+        header["Content-Type"]= "application/x-amz-json-1.1"
```
header["Content-Type"] = "application/x-amz-json-1.1"
header["X-Amz-Target"] = "AWSCognitoIdentityProviderService.InitiateAuth"

zachaubin commented on clientApp/CANLoggerClient.py:
```diff
@@ -0,0 +1,664 @@
+                self.download_file()
+
+            r1 = requests.post( response_dict['upload_link']['url'], 
+                                data=response_dict['upload_link']['fields'], 
```
data = res... 
or leave since it's param

zachaubin reviewed (commented):
See in-line comments.

jeremy-daily committed:
Added a function to list files

duyvan1995 committed:
Update requirements.txt

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Fix SHA256

Sha256 variables in codes are mixed up, should be working now

duyvan1995 committed:
Teensy Provisioning for CAN Logger 3

Not fully functional script yet, need to finish the data exchange between the Teensy and the python on line 79 and 115

duyvan1995 committed:
Update writePovisionConfig() in the library

Correct typo, add more description

duyvan1995 committed:
Update SHA256 test

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Provisioning Progress

Python now receives serial and public key from teensy

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Provision Progress #2

Logger now receives public key from AWS and load into the ATECC

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

jeremy-daily committed:
Removed Libaries

Look for the libraries in the SystemsCyber Forked branches

duyvan1995 committed:
Add AES-128 CBC test against NIST test vectors

duyvan1995 committed:
Provisioning Update

Add message dialog for operator to visually comfirm keys and execute the writing of server public key to the device

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Provisioning Update #2

Add two columns in AWS DynamoDB CAN Logger table for operator visual confirmation

duyvan1995 committed:
Update provision.py

Change len to fix parameters error code

jeremy-daily committed:
Stored Hash of Keys for verification

duyvan1995 committed:
Update sha256 test

duyvan1995 committed:
Provisioning Progress #3

Add server pem key and its encrypted password in AWS provision script to send back to user GUI

jeremy-daily committed:
encrypt private keys

duyvan1995 committed:
Provisioning hash key check

duyvan1995 committed:
Prov Hash Check

duyvan1995 committed:
Revert "Prov Hash Check"

This reverts commit 4401abb22c1b874d0242e10d0b5a07c4e5253325.

duyvan1995 committed:
Provisioning private key and password json file

duyvan1995 committed:
Provisioning: Decrypt password, clean up message box

duyvan1995 committed:
Provisioning: Update key hash display on GUI

duyvan1995 committed:
Decrypt password display

duyvan1995 committed:
Update AES functions to fit updated sparkfun library

duyvan1995 committed:
Update GUI icons

duyvan1995 committed:
Complete Format SD card function

duyvan1995 committed:
Remvoe user_input_data for upload function testing

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Add User Input Form for Uploading function

duyvan1995 committed:
Fix empty string in user input

duyvan1995 committed:
Fix public key data type

jeremy-daily committed:
Added additional error message

jeremy-daily committed:
Merge branch 'serverless' of github.com:SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Update Year in User Input

jeremy-daily committed:
Merge branch 'serverless' of github.com:SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
get_session_key and upload_user_input

Remove unnecessary and non-exist items in post data within get_session_key(). Simplify upload_user_input()

jeremy-daily committed:
Returns Session Key

jeremy-daily committed:
Merge branch 'serverless' of github.com:SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Add save decrypted file option after get session key

duyvan1995 committed:
Add log file verify during uploading

duyvan1995 committed:
Update README in serverless to deploy in Linux

duyvan1995 committed:
Add verify_upload function

duyvan1995 committed:
Fix Indentation

duyvan1995 committed:
Verify log file integrity after uploading

duyvan1995 committed:
Add upload time and verify status after succesful upload

duyvan1995 committed:
Clean up GUI menu and toolbar

Add server function buttons

duyvan1995 committed:
Fix file is not downloaded from device every upload

duyvan1995 committed:
Retrieve metadata dictionary from server

duyvan1995 committed:
Add displaying table of file metadata retrieved from server

duyvan1995 committed:
Add download file from server function

duyvan1995 committed:
Fix data type issue for downloading function

duyvan1995 committed:
Add saving file to local pc after download from server

duyvan1995 committed:
Add share access function

duyvan1995 committed:
Adding and removing email from access list in server script

duyvan1995 committed:
Fix response statement from server in share access function

duyvan1995 committed:
List server file function displays share access, revoke share function on share file

duyvan1995 committed:
Update README

duyvan1995 committed:
Fix Qemssagebox from popping up if hit cancel or close

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Add Read Info function that displays user note and access list

duyvan1995 committed:
Add download log to track who and when

duyvan1995 committed:
Change email is not case sensitive, fix access_tuple to string

duyvan1995 committed:
Update GUI

Take provision function button off tool menu, add pop up info for that provision function. Clean up read info function

jeremy-daily committed:
Update .gitignore

duyvan1995 committed:
Provision error message

if keys don't match, display error message to tell the provisoner to delete device info from CANLoggers table and provision again since the server public key has not been stored and locked in the device

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Upload function check for invalid session key

duyvan1995 committed:
Add jeremy.daily@colostate.edu as a default in access list

duyvan1995 committed:
Add default access list saving option

duyvan1995 committed:
Update access_list with email input from user in uplaod function

duyvan1995 committed:
Add applcation favicon

duyvan1995 committed:
Start command goes through setup to create new session key and IV

duyvan1995 committed:
Add access check in auth.py

duyvan1995 committed:
Add progress bar

duyvan1995 committed:
Fix Progress Bar from freezing and add stop command on logger if progress bar gets cancelled

duyvan1995 committed:
Move download function before uploading metadata

duyvan1995 committed:
Cap log file size to about 900 MB

duyvan1995 committed:
Add progress bar for loading metadata when connect to device

duyvan1995 committed:
Add more time.sleep in downloading file to prevent freezing up

duyvan1995 committed:
Remove progress bar during downloading, Add option for user to copy the file to pc

duyvan1995 committed:
Update Download Function

Add check if file exists to minimize repetition of downloading file

duyvan1995 committed:
Clean up some pop up info

duyvan1995 committed:
CLean up get_password function

duyvan1995 committed:
Fix loading progress bar with 0 file in SD card, fix file look up in auth.py

duyvan1995 committed:
Fix typo and add icon on some Qmessagebox

duyvan1995 committed:
typo

Dr-Daily committed:
Added presigned URLs for download

Dr-Daily committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

duyvan1995 committed:
Remove server private key password in the database

duyvan1995 committed:
Update downloading large file on GUI side

duyvan1995 committed:
Add close timing test, fix button pinout for new version

duyvan1995 committed:
Fix power loss problem

duyvan1995 committed:
Clean up and add description for power loss section

duyvan1995 committed:
Change prefix to CAN3

duyvan1995 committed:
Remove unnecessary file

duyvan1995 committed:
Fix meta data writing function

duyvan1995 committed:
Update Test Scripts

duyvan1995 committed:
Add ATECC608A key config test

duyvan1995 committed:
Fix key_config test for ECDH testing

duyvan1995 committed:
Update test functions for thesis writing

duyvan1995 committed:
Update ECDH and AES test script

duyvan1995 committed:
Add bit stuffing generator script

duyvan1995 committed:
Update CAN Switch in testing scripts

duyvan1995 committed:
Merge branch 'serverless' of https://github.com/SystemsCyber/CAN-Logger-3 into serverless

jeremy-daily committed:
Added DXF of enclosure

Dr-Daily committed:
Updated Laser Art for cutting

Dr-Daily committed:
hairlines

jeremy-daily committed:
website listing of CAN Loggers

jeremy-daily committed:
Merge branch 'serverless' of github.com:SystemsCyber/CAN-Logger-3 into serverless

jeremy-daily committed:
Added Website

Bootstrap Studio and Web Front end.

jeremy-daily committed:
make work

gannaramu committed:
Fix for Extracting serial_number and device_public_key

Added logic to extract the last two indeces where the serial number and device public key are present. this is just a temporary fix which until we figure out something concrete.
